dbms/src/Operators/SharedAggregateRestorer.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <Flash/Pipeline/Schedule/Events/LoadBucketEvent.h>
+#include <Interpreters/Aggregator.h>
+#include <Operators/SharedAggregateRestorer.h>
+
+namespace DB
+{
+SharedBucketDataLoader::SharedBucketDataLoader(
+    PipelineExecutorStatus & exec_status_,
+    const BlockInputStreams & bucket_streams,
+    const String & req_id,
+    size_t max_queue_size_)
+    : exec_status(exec_status_)
+    , log(Logger::get(req_id))
+    , max_queue_size(std::max(1, max_queue_size_))
+{
+    for (const auto & bucket_stream : bucket_streams)
+        bucket_inputs.emplace_back(bucket_stream);
+    assert(!bucket_inputs.empty());
+
+    exec_status.onEventSchedule();
+}
+
+SharedBucketDataLoader::~SharedBucketDataLoader()
+{
+    bucket_data_queue = {};
+    bucket_inputs.clear();
+    // In order to ensure that `PipelineExecutorStatus` will not be destructed before `SharedBucketDataLoader` is destructed.
+    exec_status.onEventFinish();
+}
+
+bool SharedBucketDataLoader::switchStatus(SharedLoaderStatus from, SharedLoaderStatus to)
+{
+    return status.compare_exchange_strong(from, to);
+}
+
+std::vector<BucketInput *> SharedBucketDataLoader::getNeedLoadInputs()
+{
+    assert(!bucket_inputs.empty());
+    std::vector<BucketInput *> load_inputs;
+    for (auto & bucket_input : bucket_inputs)
+    {
+        if (bucket_input.needLoad())
+            load_inputs.push_back(&bucket_input);
+    }
+    assert(!load_inputs.empty());
+    return load_inputs;
+}
+
+void SharedBucketDataLoader::storeBucketData()
```

Comment:
Getting spilled data from files, this my understanding of the purpose of this function. So why we name this function as `storexxx`, why not `restorexxx`?

Reply:
Because the blocks read here will be stored in the `SharedBucketDataLoader::bucket_data_queue`.